### PR TITLE
[samesite] adding cluster version verification for samesite cookie support

### PR DESCRIPTION
**Pull request by treymer**

### CODE REVIEW

File: pkg/e2e/verify/samesite_cookie.go
Comment (whereswaldon) [inline pkg/e2e/verify/samesite_cookie.go:39]: Since this exact same code block is repeated twice, would it be possible to refactor it into a single function. Might work well as a closure defined above the `ginkgo.It` calls...

File: pkg/e2e/verify/samesite_cookie.go
Comment (whereswaldon) [inline pkg/e2e/verify/samesite_cookie.go:76]: Would you consider adding this as a method to `helper.H` instead and exporting it? I could see other `osde2e` users wanting to be able to do this.

**Review by whereswaldon — CHANGES_REQUESTED** [review]
Looks solid, well done! I think there are some small opportunities to DRY it up and make other testers' lives easier that I'd like to talk about before we call it done.

File: pkg/e2e/verify/samesite_cookie.go
Comment (treymer) [reply pkg/e2e/verify/samesite_cookie.go:39]: Yeah, I originally attempted to create a function and ginkgo was skipping the second test for some reason. Closures are new to me, I will check that out and test.

File: pkg/e2e/verify/samesite_cookie.go
Comment (treymer) [reply pkg/e2e/verify/samesite_cookie.go:76]: Absolutely, that is a great idea. I will explore this.

**Comment by jeefy** [comment]
/approve
Strong +1 to this. Super excited to also see adding support for other tests to consume it as well!

**Comment by treymer** [comment]
Put this test back into informing for now since it is a major functional change to this test I want to ensure it bakes there for a couple days and then I will promote back to e2e.

**Comment by treymer** [comment]
@whereswaldon @jeefy Added GetClusterVersion to helper and DRY'd up my code w/ a closure function.

File: pkg/common/helper/helper.go
Comment (whereswaldon) [inline pkg/common/helper/helper.go:353]: I think this might be kinda pedantic and awful of me, but I don't think I understand why it's worth adding another return value here when you're already returning the struct containing that value as the first return value. I can kinda see that the string of nested field accesses is long and it saves you from doing that in other code, but I'm not sure that I think the tradeoff is worth it.

I won't block on this, but I'd recommend dropping back to two return values.

File: pkg/e2e/verify/samesite_cookie.go
Comment (whereswaldon) [inline pkg/e2e/verify/samesite_cookie.go:61]: It's always been this way, so I guess I really should have paid more attention last time, but this can panic if the current version isn't what we expect (a semver). Is there any realistic chance of that happening? Feels like we should defensively do something like:
```suggestion
		if len(splitVersion) < 2 {
			return true // not supported
		}
		majMinVersion, err := strconv.Atoi(splitVersion[0] + splitVersion[1])
```

**Review by whereswaldon — COMMENTED** [review]
Thanks for that refactor! I have two more thoughts on this codebase. You don't need to change anything if you're happy with how things stand, but I'd at least like to hear what you think before giving my approval.

File: pkg/common/helper/helper.go
Comment (treymer) [reply pkg/common/helper/helper.go:353]: Fair enough, keeping the business logic down stream makes sense due to unintended increases in time complexity. I will adjust :)

File: pkg/e2e/verify/samesite_cookie.go
Comment (treymer) [reply pkg/e2e/verify/samesite_cookie.go:61]: While I think it would be unlikely, I am new to the team in general so I am unsure if a non semver could make its way into the API. Either way, likely worth the change to add a catch. :+1:

**Comment by treymer** [comment]
> Thanks for that refactor! I have two more thoughts on this codebase. You don't need to change anything if you're happy with how things stand, but I'd at least like to hear what you think before giving my approval.

I agree with the suggested changes. I will make the changes and push a new commit here shortly :smile:

**Comment by treymer** [comment]
@whereswaldon changes have been made and tested. Let me know what you think! :smile:

File: pkg/e2e/verify/samesite_cookie.go
Comment (whereswaldon) [inline pkg/e2e/verify/samesite_cookie.go:63]: I don't think this handles the (unlikely) case of len == {1,2,3}. 3 and 2 both don't cause panics, but 1 does.

**Comment by whereswaldon** [comment]
Mostly this looks great! I think there's still an edge case in one place though.

File: pkg/e2e/verify/samesite_cookie.go
Comment (treymer) [reply pkg/e2e/verify/samesite_cookie.go:63]: Good point, let me make it simpler.

Comment (treymer) [reply pkg/e2e/verify/samesite_cookie.go:63]: @whereswaldon I can just do this:
`		if len(splitVersion) != 2 || len(splitVersion) != 3 {
			return true
		}`
This way we explicitly expect a len or 2 or 3, anything else is unsupported.

**Comment by treymer** [comment]
@whereswaldon This change should account for {0,1} and >3 so we should always expect either {2,3}. Let me know what you think. Thanks!

File: pkg/e2e/verify/samesite_cookie.go
Comment (whereswaldon) [inline pkg/e2e/verify/samesite_cookie.go:79]: So if `supportVersion` is `46`, and we have a cluster at version `4.7.11`, we end up comparing `4711 < 46`. Is that the intention? In general, integer comparison here might be problematic once the minor/patch versions have multiple digits? I don't know the constraints that you're working under here, so perhaps there's a reason that you're not worried about that?

I guess I should ask: What is the goal of this function? What does it verify about the version? What does a true or false return value *mean*?

Comment (treymer) [reply pkg/e2e/verify/samesite_cookie.go:79]: SameSite cookie being enables on routes in only supported in OSD versions >4.6. So, I am trying to verify the version here. In cases for version 5.11 for example I want 511 < 46 to fail. That is the intention here. Comparing integers is the way I am able to use conditionals in this sense to verify versions. I cannot simply look for versions >4.6 because 5.11 won't pass. So, this is the only way I can think of doing it. If you have other suggestions, then I am all ears.

This is why I check the length to be either 2 or 3 to account for multiple digits in the minor version (5.11) for example. Also, keep in mind I am splitting off the patch version, because I do not care about that in this case.

Comment (treymer) [reply pkg/e2e/verify/samesite_cookie.go:79]: I actually do not need the conditional to check if len is 3, so I will remove that.

Comment (treymer) [reply pkg/e2e/verify/samesite_cookie.go:79]: I can clean it up to this:
`if len(splitVersion) == 2 {
			majMinVersion, err := strconv.Atoi(splitVersion[0] + splitVersion[1])
			Expect(err).NotTo(HaveOccurred(), "failed normalizing major/minor version %v to integer %v", majMinVersion, err)
			if majMinVersion < supportVersion {
				return true
			} else {
				return false
			}
		}`

File: pkg/e2e/verify/samesite_cookie.go
Comment (treymer) [reply pkg/e2e/verify/samesite_cookie.go]: You beat me to it! :laughing:

**Comment by treymer** [comment]
@whereswaldon ready for your review. I tested against all use cases discussed here.

**Review by whereswaldon — COMMENTED** [review]
/lgtm

**Pull request merged**
